### PR TITLE
A line item matcher

### PR DESCRIPTION
**Pull request by dwkitchen**
Is this the basics of what you are after?

### CODE REVIEW

**Comment by dwkitchen** [comment]
What about customisable line items? I guess this is handled with `drupal_alter('commerce_cart_product_comparison_properties', $comparison_properties, clone($line_item));` in Commerce 1.x

How will customisations be handled by the Purchasable Entity?

**Comment by dwkitchen** [comment]
I still don't get where a customisable product would have its customisations. It doesn't look like they are in the PurchasableEntity, so I think we need a second variable to this matcher if a dummy line item is not created.

File: modules/order/src/LineItemMatcher.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+   */
+  protected $order;
+
+  /**
+   * Creates a LineItemMatcherBase instance.
+   *
+   * @param OrderInterface $order
+   */
+  public function __construct($order)
+  {
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php]: Opening brace on wrong position.

File: modules/order/src/LineItemMatcher.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  /**
+   * The order.
+   *
+   * @var Drupal\commerce_order\Entity\OrderInterface
+   */
+  protected $order;
+
+  /**
+   * Creates a LineItemMatcherBase instance.
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php]: Constructs a new $X object.

File: modules/order/src/LineItemMatcher.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   *
+   * @param OrderInterface $order
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php]: Param must have full namespace, description in the line below.

File: modules/order/src/LineItemMatcher.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+namespace Drupal\commerce_order;
+
+use Drupal\commerce\PurchasableEntityInterface;
+use Drupal\commerce_order\Entity\OrderInterface;
+use Drupal\commerce_order\Entity\LineItemInterface;
+use Drupal\commerce_order\Event\OrderEvents;
+use Drupal\commerce_order\Event\LineItemComparisonFieldsEvent;
+
+/**
+ * Class LineItemMatcher
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php:17]: Missing docs.

File: modules/order/src/Event/LineItemComparisonFieldsEvent.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  /**
+   * Constructs a new LineItemComparisonFieldsEvent object.
+   *
+   * @param \Drupal\commerce\PurchasableEntityInterface $entity
+   *   The Purchasable Entity
+   */
+  public function __construct(PurchasableEntityInterface $entity) {
+    $this->entity = $entity;
+    $this->comparisonFields = ['type', 'id'];
```
Comment (bojanz) [inline modules/order/src/Event/LineItemComparisonFieldsEvent.php]: We should pass these to the constructor, to keep the logic in the Matcher.

File: modules/order/src/LineItemMatcher.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function matchAll(PurchasableEntityInterface $entity, OrderInterface $order) {
+
+    // Get the comparison fields from all modules altering them.
+    $event = new LineItemComparisonFieldsEvent($entity);
+    $event_dispatcher = \Drupal::service('event_dispatcher');
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php]: Inject the event dispatcher.

Comment (dwkitchen) [inline modules/order/src/LineItemMatcher.php]: This is how it is done in the address module. How should it be done?

File: modules/order/src/LineItemMatcher.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ * @package Drupal\commerce_order
+ */
+class LineItemMatcher implements LineItemMatcherInterface {
+
+  /**
+   * An Event Dispatcher.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $event_dispatcher;
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php:30]: Class attributes must be camelCase.

File: modules/order/src/LineItemMatcher.php [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ * Class LineItemMatcher
+ *
+ * Match an existing line item for the PurchasableEntity.
+ *
+ * @package Drupal\commerce_order
+ */
+class LineItemMatcher implements LineItemMatcherInterface {
+
+  /**
+   * An Event Dispatcher.
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php:26]: "The event dispatcher."

File: modules/order/src/LineItemMatcher.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $event_dispatcher;
+
+  /**
+   * Constructs a new LineItemMatcher.
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php:33]: Constructs a new X object.

File: modules/order/src/LineItemMatcher.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+   *
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $event_dispatcher
```
Comment (bojanz) [inline modules/order/src/LineItemMatcher.php:35]: Missing description below param ("The event dispatcher.")

**Comment by bojanz** [comment]
Committed a more advanced version of this in d25878624dd27bc0f7090102a1f37e1e1399cc38.

**Pull request closed without merging**